plugins/outputs/sematext/sematext.go

```diff
@@ -0,0 +1,294 @@
+
+		return s.handleResponse(res)
+	}
+
+	return nil
+}
+
+func (s *Sematext) handleResponse(res *http.Response) error {
+	success, badRequest := checkResponseStatus(res)
+
+	if !success {
+		errorMsg := fmt.Sprintf("received %d status code, message = %q while sending to %s",
+			res.StatusCode, res.Status, s.metricsURL)
+
+		if badRequest {
+			// shouldn't be re-sent as bad request will continue to be a bad request
+			s.Log.Errorf("%s - request will be dropped", errorMsg)
+			return nil
+		}
+
+		// otherwise it is some temporary error from the backend and we should retry
+		return fmt.Errorf(errorMsg)
+	}
+
+	return nil
+}
+
+func checkResponseStatus(res *http.Response) (success bool, badRequest bool) {
+	success = res.StatusCode >= 200 && res.StatusCode < 300
+	badRequest = res.StatusCode >= 400 && res.StatusCode < 500
+
+	return success, badRequest
+}
+
+// processMetrics returns an error only when the whole batch of metrics should be discarded
+// batchProcessors run first, metricProcessors follow later
+func (s *Sematext) processMetrics(metrics []telegraf.Metric) ([]telegraf.Metric, error) {
+	s.Log.Debugf("Starting processing of slice of %d metrics", len(metrics))
+
+	if metricsAlreadyProcessed(metrics) {
+		// in case some batch was fully processed before, we don't want to process it once again
+		s.Log.Debugf("Skipping processing of already processed slice of %d metrics", len(metrics))
+		return metrics, nil
+	}
+
+	for _, p := range s.batchProcessors {
+		var err error
+		metrics, err = p.Process(metrics)
+
+		if err != nil {
+			s.Log.Errorf("error while running batch processors in Sematext output: %v", err)
+			return metrics, err
+		}
+	}
+
+	processedMetrics := make([]telegraf.Metric, 0, len(metrics))
+
+	for _, metric := range metrics {
+		metricOk := true
+
+		// don't process the metrics that were already processed before
+		if !metricAlreadyProcessed(metric) {
+			for _, p := range s.metricProcessors {
+				err := p.Process(metric)
+
+				if err != nil {
+					// log the message, mark the metric to be skipped, skip other processors
+					s.Log.Warnf("can't process metric: %s in Sematext output, error : %s", metric, err.Error())
+					metricOk = false
+					break
+				}
+			}
+		}
+
+		if metricOk {
+			processedMetrics = append(processedMetrics, metric)
+		}
+	}
+
+	markMetricsProcessed(processedMetrics)
+
+	return processedMetrics, nil
+}
+
+func metricsAlreadyProcessed(metrics []telegraf.Metric) bool {
+	// return that batch hasn't been processed yet if any of its metrics hasn't been processed
+	for _, m := range metrics {
+		if !metricAlreadyProcessed(m) {
+			return false
+		}
+	}
+
+	return true
+}
+
+func metricAlreadyProcessed(metric telegraf.Metric) bool {
+	_, processed := metric.GetTag(tags.SematextProcessedTag)
+	return processed
+}
+
+func markMetricsProcessed(metrics []telegraf.Metric) {
+	for _, m := range metrics {
+		m.AddTag(tags.SematextProcessedTag, tags.SematextProcessedTag)
+	}
+}
```

Comment:
I think using the metrics to track the sending state is completely flawed. Once you processed them they are dropped anyway, so why do you need to mark them?